test/Autofac.Test/Core/ReflectionCacheSetTests.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Autofac Project. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Reflection;
+using Autofac.Core;
+using Autofac.Util.Cache;
+
+namespace Autofac.Test.Core;
+
+public class ReflectionCacheSetTests
+{
+    [Fact]
+    public void ClearingReflectionCacheBetweenResolvesIsOk()
+    {
+        var builder = new ContainerBuilder();
+
+        builder.RegisterType<CDerivedSingle<int>>().As<ISingle<int>>();
+
+        var container = builder.Build();
+
+        container.Resolve<ISingle<int>>();
+
+        ReflectionCacheSet.Shared.Clear();
+
+        container.Resolve<ISingle<int>>();
+    }
+
+    [Fact]
+    public async Task ConcurrentCacheClearsAndResolvesIsOk()
+    {
+        var builder = new ContainerBuilder();
```

Comment:
I wonder if we're going to run into other weirdness with testing shared cache clearing. It's a `static` member, and while other tests _in this fixture_ won't run in parallel, [other test fixtures _may_ run in parallel](https://xunit.net/docs/running-tests-in-parallel), thus making it possible that different tests will execute the clear on build or something like that due to the order of `ContainerBuilder` creation. I don't think (?) there's isolation around static references across test fixtures.

Does this class need a `[CollectionDefinition(DisableParallelization = true)]` attribute to ensure nothing else runs at the same time?

Reply:
Possibly related - is there a need, from a testing perspective, to be able to "reset" the `IsFirstContainerBuilder` value?

Reply:
I believe the only way to ensure those tests run in isolation is to have them in a separate project; `CollectionDefinition` has to be applied to a shared collection class referenced by all classes that need to hold off while the resource is in use, which isn't workable for obvious reasons.

One comment I might make though upon reflection is...what am I actually testing here with these concurrency tests? It feels like I'm just testing that the `ConcurrentDictionary.GetOrAdd` function works on multiple threads? Which seems unnecessary; other code is testing that existing functionality that uses the `GetOrCreateCache` method continues to work, and lord knows the dotnet team has plenty of tests for the `ConcurrentDictionary`.

In a sense it may be better to test the `ReflectionCacheSet` without using the `Shared` instance, therefore conducting an actual unit test rather than a strange concurrency integration test that I thought was a good idea for some reason.

Reply:
I was under the impression that each test class was implicitly its own collection, but I admit I haven't really messed with it at this level. I've had to disable parallelism for a whole assembly in a few cases, but never at a specific fixture level. But the point that maybe "these aren't the tests you're looking for" is valid. /obi-wan

Reply:
OK, have updated. Hopefully these _are_ the tests I am looking for /stormtrooper